File: mhycli/bili_sdk.py
"""B站崩坏3 (B服) 登录 — 对应开源版 BSGameSDK.hpp

流程:
  1. 请求 /api/client/rsa 获取 RSA 公钥, 加密密码
  2. POST /api/client/login 登录 → access_key + uid
  3. POST /api/client/user.info 获取昵称
  (若需极验验证码: 先 /api/client/start_captcha 获取 gt/challenge)
"""
from __future__ import annotations

import json
import time

import requests

from .crypto import md5, rsa_encrypt_base64

BILI_BASE = "https://line1-sdk-center-login-sh.biligame.net"
BILI_RSA = f"{BILI_BASE}/api/client/rsa"
BILI_LOGIN = f"{BILI_BASE}/api/client/login"
BILI_USERINFO = f"{BILI_BASE}/api/client/user.info"
BILI_CAPTCHA = f"{BILI_BASE}/api/client/start_captcha"

HEADERS = {
    "User-Agent": "Mozilla/5.0 BSGameSDK",
    "Content-Type": "application/x-www-form-urlencoded",
    "Host": "line1-sdk-center-login-sh.biligame.net",
}

# BSGameSDK.hpp 中写死的设备参数 (MuMu 模拟器指纹)
BASE_PARAM = {
    "operators": "5", "merchant_id": "590", "isRoot": "0", "domain_switch_count": "0",
    "sdk_type": "1", "sdk_log_type": "1",
    "support_abis": "x86,armeabi-v7a,armeabi",
    "access_key": "", "sdk_ver": "3.4.2", "oaid": "", "dp": "1280*720",
    "original_domain": "", "imei": "",
    "version": "1", "udid": "KREhESMUIhUjFnJKNko2TDQFYlZkB3cdeQ==",
    "apk_sign": "4502a02a00395dec05a4134ad593224d", "platform_type": "3",
    "old_buvid": "XZA2FA4AC240F665E2F27F603ABF98C615C29",
    "android_id": "84567e2dda72d1d4", "fingerprint": "", "mac": "08:00:27:53:DD:12",
    "server_id": "378", "domain": "line1-sdk-center-login-sh.biligame.net",
    "app_id": "180", "version_code": "510", "net": "4", "pf_ver": "12",
    "cur_buvid": "XZA2FA4AC240F665E2F27F603ABF98C615C29", "c": "1",
    "brand": "Android", "channel_id": "1", "uid": "", "game_id": "180",
    "ver": "6.1.0", "model": "MuMu",
}


def _set_sign(data: dict) -> str:
    """对应 BSGameSDK::detail::SetSign
    除 pwd 外不 urlEncode, 生成 k=v& 串, 再加 sign=md5(所有值拼接+key)
    """
    data = dict(data)
    data["timestamp"] = str(int(time.time()))
    data["client_timestamp"] = str(int(time.time()))
    from urllib.parse import quote

    body_parts = []
    values = []
    for key, value in data.items():
        if isinstance(value, str):
            sv = value
        else:
            sv = json.dumps(value, ensure_ascii=False, separators=(",", ":"))
        if key == "pwd":
            body_parts.append(f"{key}={quote(sv)}")
        else:
            body_parts.append(f"{key}={sv}")
        values.append(sv)
    sign_raw = "".join(values) + "dbf8f1b4496f430b8a3c0f436a35b931"
    body = "&".join(body_parts) + "&" + "sign=" + md5(sign_raw)
    return body


def _encrypted_pwd(password: str) -> str:
    """对应 BSGameSDK::detail::GetEncryptedPwd
    先请求 rsa 接口拿公钥+hash, 再对 hash+pwd 做 RSA 加密
    """
    rsa_param = {k: v for k, v in BASE_PARAM.items() if k not in ("access_key",)}
    rsa_param["access_key"] = ""
    rsa_param["timestamp"] = str(int(time.time()))
    rsa_param["client_timestamp"] = str(int(time.time()))
    body = _set_sign(rsa_param)
    resp = requests.post(BILI_RSA, data=body, headers=HEADERS)
    resp.raise_for_status()
    info = resp.json()
    public_key = info["rsa_key"]
    hash_ = info["hash"]
    return rsa_encrypt_base64(hash_ + password, public_key)


def login_by_password(account: str, password: str, gt_user: str = "",
                      challenge: str = "", validate: str = "") -> dict:
    """对应 BSGameSDK::LoginByPassWord
    返回 {code, message, uid, access_key, uname}
    """
    data = {k: v for k, v in BASE_PARAM.items()}
    data.update({
        "access_key": "", "gt_user_id": gt_user, "uid": "",
        "challenge": challenge, "user_id": account, "validate": validate,
    })
    if validate:
        data["seccode"] = validate + "|jordan"
    data["pwd"] = _encrypted_pwd(password)

    body = _set_sign(data)
    resp = requests.post(BILI_LOGIN, data=body, headers=HEADERS)
    resp.raise_for_status()
    info = resp.json()
    result = {"code": info.get("code", -1), "message": info.get("message", ""),
              "uid": "", "access_key": "", "uname": ""}
    if info.get("code") in (0,):
        result["uid"] = str(info.get("uid", ""))
        result["access_key"] = info.get("access_key", "")
        ui = get_user_info(result["uid"], result["access_key"])
        result["uname"] = ui["uname"]
    return result


def get_user_info(uid: str, access_key: str) -> dict:
    """对应 BSGameSDK::GetUserInfo"""
    data = {k: v for k, v in BASE_PARAM.items()}
    data.update({"uid": uid, "access_key": access_key})
    body = _set_sign(data)
    resp = requests.post(BILI_USERINFO, data=body, headers=HEADERS)
    resp.raise_for_status()
    info = resp.json()
    return {"code": info.get("code", -1), "uname": info.get("uname", "")}


def start_captcha() -> dict:
    """对应 BSGameSDK::CaptchaCaptcha — 返回 GeetestData"""
    data = {k: v for k, v in BASE_PARAM.items()}
    body = _set_sign(data)
    resp = requests.post(BILI_CAPTCHA, data=body, headers=HEADERS)
    resp.raise_for_status()
    cap = resp.json()
    return {"gt": cap.get("gt", ""), "challenge": cap.get("challenge", ""),
            "gt_user_id": cap.get("gt_user_id", "")}

File: src/bili_sdk.rs
//! B站崩坏3（B服）账号密码登录，对应 Python `bili_sdk.py` / BSGameSDK.hpp。
//!
//! 流程：`/api/client/rsa` 取公钥 → RSA 加密 `hash+pwd` → `/api/client/login`
//! → `/api/client/user.info` 取昵称；需要极验时先 `/api/client/start_captcha`。
use anyhow::{Result, bail};
use base64::Engine as _;
use base64::engine::general_purpose::STANDARD;
use percent_encoding::{NON_ALPHANUMERIC, utf8_percent_encode};
use reqwest::header::{CONTENT_TYPE, USER_AGENT};
use rsa::pkcs8::DecodePublicKey;
use rsa::{Pkcs1v15Encrypt, RsaPublicKey};
use serde_json::Value;

use crate::crypto::md5_hex;

const BILI_BASE: &str = "https://line1-sdk-center-login-sh.biligame.net";
const BILI_RSA: &str = "https://line1-sdk-center-login-sh.biligame.net/api/client/rsa";
const BILI_LOGIN: &str = "https://line1-sdk-center-login-sh.biligame.net/api/client/login";
const BILI_USERINFO: &str = "https://line1-sdk-center-login-sh.biligame.net/api/client/user.info";
const BILI_CAPTCHA: &str =
    "https://line1-sdk-center-login-sh.biligame.net/api/client/start_captcha";

const UA: &str = "Mozilla/5.0 BSGameSDK";
const BILI_APP_KEY: &str = "dbf8f1b4496f430b8a3c0f436a35b931";

/// 有序参数表（保持插入顺序，更新已存在键不改变位置）
#[derive(Clone)]
struct Params(Vec<(String, String)>);

impl Params {
    fn base() -> Self {
        let pairs: &[(&str, &str)] = &[
            ("operators", "5"),
            ("merchant_id", "590"),
            ("isRoot", "0"),
            ("domain_switch_count", "0"),
            ("sdk_type", "1"),
            ("sdk_log_type", "1"),
            ("support_abis", "x86,armeabi-v7a,armeabi"),
            ("access_key", ""),
            ("sdk_ver", "3.4.2"),
            ("oaid", ""),
            ("dp", "1280 * 720"),
            ("original_domain", ""),
            ("imei", ""),
            ("version", "1"),
            ("udid", "XXA31CBAB6CBA63E432E087B58411A213BFB7"),
            ("apk_sign", "4502a02a00395dec05a4134ad593224d"),
            ("platform_type", "3"),
            ("old_buvid", "XZA2FA4AC240F665E2F27F603ABF98C615C29"),
            ("android_id", "84567e2dda72d1d4"),
            ("fingerprint", ""),
            ("mac", "08:00:27:53:DD:12"),
            ("server_id", "378"),
            ("domain", "line1-sdk-center-login-sh.biligame.net"),
            ("app_id", "180"),
            ("version_code", "510"),
            ("net", "4"),
            ("pf_ver", "12"),
            ("cur_buvid", "XZA2FA4AC240F665E2F27F603ABF98C615C29"),
            ("c", "1"),
            ("brand", "Android"),
            ("channel_id", "1"),
            ("uid", ""),
            ("game_id", "180"),
            ("ver", "6.1.0"),
            ("model", "MuMu"),
        ];
        Params(
            pairs
                .iter()
                .map(|(k, v)| (k.to_string(), v.to_string()))
                .collect(),
        )
    }

    fn set(&mut self, k: &str, v: impl Into<String>) {
        let v = v.into();
        if let Some(slot) = self.0.iter_mut().find(|(key, _)| key == k) {
            slot.1 = v;
        } else {
            self.0.push((k.to_string(), v));
        }
    }

    fn remove(&mut self, k: &str) {
        self.0.retain(|(key, _)| key != k);
    }
}

/// 对应 BSGameSDK::detail::SetSign（时间戳为毫秒）
fn set_sign(mut data: Params) -> String {
    let ts = crate::crypto::now_millis().to_string();
    data.set("timestamp", ts.clone());
    data.set("client_timestamp", ts);

    let mut values = String::new();
    let mut parts = Vec::new();
    for (k, v) in &data.0 {
        values.push_str(v);
        if k == "pwd" {
            parts.push(format!("{k}={}", utf8_percent_encode(v, NON_ALPHANUMERIC)));
        } else {
            parts.push(format!("{k}={v}"));
        }
    }
    let sign = md5_hex(&format!("{values}{BILI_APP_KEY}"));
    format!("{}&sign={sign}", parts.join("&"))
}

async fn post(http: &reqwest::Client, url: &str, body: String) -> Result<Value> {
    let resp = http
        .post(url)
        .header(USER_AGENT, UA)
        .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
        .body(body)
        .send()
        .await?;
    Ok(resp.json().await?)
}

/// RSA PKCS#1 v1.5 加密 + Base64
fn rsa_encrypt_base64(message: &str, public_key_pem: &str) -> Result<String> {
    let key = RsaPublicKey::from_public_key_pem(public_key_pem)
        .map_err(|e| anyhow::anyhow!("解析公钥失败: {e}"))?;
    let mut rng = rand::thread_rng();
    let enc = key
        .encrypt(&mut rng, Pkcs1v15Encrypt, message.as_bytes())
        .map_err(|e| anyhow::anyhow!("RSA 加密失败: {e}"))?;
    Ok(STANDARD.encode(enc))
}

/// 对应 BSGameSDK::detail::GetEncryptedPwd
async fn encrypted_pwd(http: &reqwest::Client, password: &str) -> Result<String> {
    let mut rsa_param = Params::base();
    for k in [
        "uid",
        "pwd",
        "challenge",
        "validate",
        "seccode",
        "gt_user_id",
        "user_id",
        "captcha_type",
    ] {
        rsa_param.remove(k);
    }
    let body = set_sign(rsa_param);
    let info = post(http, BILI_RSA, body).await?;
    let public_key = info.get("rsa_key").and_then(Value::as_str).unwrap_or("");
    let hash = info.get("hash").and_then(Value::as_str).unwrap_or("");
    if public_key.is_empty() || hash.is_empty() {
        bail!("rsa 接口返回异常: {info}");
    }
    rsa_encrypt_base64(&format!("{hash}{password}"), public_key)
}

pub struct BiliLoginResult {
    pub code: i64,
    pub message: String,
    pub uid: String,
    pub access_key: String,
    pub uname: String,
}

/// 账号密码登录（可带极验 challenge/validate/gt_user_id）
pub async fn login_by_password(
    http: &reqwest::Client,
    account: &str,
    password: &str,
    gt_user: &str,
    challenge: &str,
    validate: &str,
) -> Result<BiliLoginResult> {
    let mut data = Params::base();
    data.set("access_key", "");
    data.set("gt_user_id", gt_user);
    data.set("uid", "");
    data.set("challenge", challenge);
    data.set("user_id", account);
    data.set("validate", validate);
    data.set(
        "seccode",
        if validate.is_empty() {
            String::new()
        } else {
            format!("{validate}|jordan")
        },
    );
    let pwd = encrypted_pwd(http, password).await?;
    data.set("pwd", pwd);
    data.set("captcha_type", "1");

    let info = post(http, BILI_LOGIN, set_sign(data)).await?;
    let code = info.get("code").and_then(Value::as_i64).unwrap_or(-1);
    let mut result = BiliLoginResult {
        code,
        message: info
            .get("message")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        uid: String::new(),
        access_key: String::new(),
        uname: String::new(),
    };
    if code == 0 {
        result.uid = info.get("uid").map(scalar).unwrap_or_default();
        result.access_key = info
            .get("access_key")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        if let Ok(ui) = get_user_info(http, &result.uid, &result.access_key).await {
            result.uname = ui;
        }
    }
    Ok(result)
}

/// 对应 BSGameSDK::GetUserInfo，返回昵称
pub async fn get_user_info(http: &reqwest::Client, _uid: &str, access_key: &str) -> Result<String> {
    let mut data = Params::base();
    data.set("uid", "");
    data.set("access_key", access_key);
    let info = post(http, BILI_USERINFO, set_sign(data)).await?;
    Ok(info
        .get("uname")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string())
}

/// 对应 BSGameSDK::CaptchaCaptcha，返回 (gt, challenge, gt_user_id)
pub async fn start_captcha(http: &reqwest::Client) -> Result<(String, String, String)> {
    let info = post(http, BILI_CAPTCHA, set_sign(Params::base())).await?;
    Ok((
        info.get("gt")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        info.get("challenge")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        info.get("gt_user_id")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
    ))
}

fn scalar(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

/// 基础域名（供上层日志展示）
pub fn base_url() -> &'static str {
    BILI_BASE
}
